rust/src/dirdiff.rs

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (C) Red Hat, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ */
+
+use anyhow::{Context, Result};
+use openat_ext::OpenatDirExt;
+use serde_derive::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::collections::BTreeSet;
+use std::convert::TryFrom;
+use std::fmt;
+
+use std::io::Read;
+
+pub(crate) type FileSet = BTreeSet<String>;
+
+/// Diff between two directories.
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub(crate) struct Diff {
+    /// Files that are new in an existing directory
+    pub(crate) added_files: FileSet,
+    /// New directories
+    pub(crate) added_dirs: FileSet,
+    /// Files removed
+    pub(crate) removed_files: FileSet,
+    /// Directories removed (recursively)
+    pub(crate) removed_dirs: FileSet,
+    /// Files that changed (in any way, metadata or content)
+    pub(crate) changed_files: FileSet,
+    /// Directories that changed mode/permissions
+    pub(crate) changed_dirs: FileSet,
+}
+
+impl Diff {
+    #[allow(dead_code)]
+    pub(crate) fn count(&self) -> usize {
+        self.added_files.len()
+            + self.added_dirs.len()
+            + self.removed_files.len()
+            + self.removed_dirs.len()
+            + self.changed_files.len()
+            + self.changed_dirs.len()
+    }
+
+    pub(crate) fn contains(&self, s: &str) -> bool {
+        self.added_files.contains(s)
+            || self.added_dirs.contains(s)
+            || self.removed_files.contains(s)
+            || self.removed_dirs.contains(s)
+            || self.changed_files.contains(s)
+            || self.changed_dirs.contains(s)
+    }
+}
+
+impl fmt::Display for Diff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "files(added:{} removed:{} changed:{}) dirs(added:{} removed:{} changed:{})",
+            self.added_files.len(),
+            self.removed_files.len(),
+            self.changed_files.len(),
+            self.added_dirs.len(),
+            self.removed_dirs.len(),
+            self.changed_dirs.len()
+        )
+    }
+}
+
+fn file_content_changed(
+    src: &openat::Dir,
+    dest: &openat::Dir,
+    path: &str,
+    expected_len: u64,
+) -> Result<bool> {
+    let mut remaining = expected_len;
+    let mut srcf = std::io::BufReader::new(src.open_file(path)?);
+    let mut destf = std::io::BufReader::new(dest.open_file(path)?);
+    let mut srcbuf = [0; 4096];
+    let mut destbuf = [0; 4096];
+    let bufsize = srcbuf.len();
+    while remaining > 0 {
+        let readlen = std::cmp::min(usize::try_from(remaining).unwrap_or(bufsize), bufsize);
+        let mut srcbuf = &mut srcbuf[0..readlen];
+        let mut destbuf = &mut destbuf[0..readlen];
+        srcf.read_exact(&mut srcbuf)?;
+        destf.read_exact(&mut destbuf)?;
+        if srcbuf != destbuf {
+            return Ok(true);
+        }
+        remaining -= readlen as u64;
```

Comment:
Can we make the underflow behavior of this explicit? My understanding is that the code around here is designed so that this never underflows, right?